In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

### Assess
In the cells below, each column of each table in this clinical trial dataset is described. To see the table that goes hand in hand with these descriptions, display each table in its entirety by displaying the pandas DataFrame that it was gathered into. This task is the mechanical part of visual assessment in pandas.

In [5]:
# display patients table
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [7]:
treatments.head(10)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   
5    jasmine       sykes          -  42u - 44u         7.56       7.18   
6     sophia      haugen  37u - 42u          -         7.65       7.27   
7      eddie      archer  31u - 38u          -         7.89       7.55   
8      saber      ménard          -  54u - 54u         8.08       7.70   
9       asia     woźniak  30u - 36u          -         7.76       7.37   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32  
5          0.38  
6          0.38  
7          0.34  
8           NaN  
9           NaN

In [8]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

## Data Quality(Dirty data)
    treatment: data missing values(NaN) in hba1c_change
    patient: zip code in patients table is float and not string
    patient: height of 'Time Neudorf' looks incorrect
    patient: state names abbreviations and full form

In [11]:
patients[patients['given_name'] == 'Tim'].index[0]

4

In [18]:
patients.iloc[8]

patient_id                                            9
assigned_sex                                       male
given_name                                        Dsvid
surname                                      Gustafsson
address                              1790 Nutter Street
city                                        Kansas City
state                                                MO
zip_code                                        64105.0
country                                   United States
contact         816-265-9578DavidGustafsson@armyspy.com
birthdate                                      3/6/1937
weight                                            163.9
height                                               66
bmi                                                26.5
Name: 8, dtype: object

In [19]:
patients[patients['given_name']=='Tám']

patient_id assigned_sex given_name surname             address    city  \
16          17       female        Tám    Liễu  2152 Heritage Road  Fresno   

         state  zip_code        country  \
16  California   93706.0  United States   

                                   contact   birthdate  weight  height   bmi  
16  LieuThiThuTam@dayrep.com1 559 765 7836  11/14/1952   183.9      61  34.7

In [22]:
treatments.info()
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   st

### Quality
    patients table
        - zip code is a float not a string
        - zip code has four digits sometimes
        - Tim Neudorf height is 27 in instead of 72 in
        - full state names sometimes, abbreviations other times
        - Dsvid Gustafsson
        - Missing demographic information(address - contact columns)
        - Erroneous dataypes(zip_code, birthdate, state(object) sex(categorical dt exists in pd))
        - multiple phone number format
        - Default John Doe data
        - Multiple records for Jakobsen, Gersten, Taylor
        
    treatments table
        - missing HbA1c changes
        - the letter u in starting and ending doses for Auralin and Novodra
        - lowercase given names and surnames
        - missing records (280 instead of 350)
        - HBa1change has 4s mistaken as 9- could be optical resolution s/w issue
        - Null represented as - in the table
        
    adverse_reactions table
        - lowercase given names and surnames

### Term	Definition
- Accuracy:	Quality dimension focused on whether the data is incorrect
- Completeness:	Quality dimension focused on whether we have all of the records we should
- Consistency:	Quality dimension focused on whether the data conforms to a standardized format
- Validity:	Quality dimension focused on whether the data conforms to a defined schema

### Programmatic assesment
    - Info method
    - sample method: provide non-directed random assessment
    

In [23]:
treatments.info()
# missing data in hba1c_change for instace

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [24]:
treatments.sample(5)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
190     martha  afanasyeva          -  34u - 41u         7.81       7.39   
232     pamela        hill          -  27u - 29u         9.53       9.10   
272     steven         roy          -  41u - 43u         7.87       7.43   
145      ruman     bisliev  46u - 53u          -         7.72       7.39   
64     cumelén  miramontes  24u - 32u          -         7.71       7.33   

     hba1c_change  
190           NaN  
232          0.93  
272          0.94  
145          0.33  
64           0.38

In [25]:
patients.head()
patients.tail()

patient_id assigned_sex given_name      surname  \
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state  zip_code  \
498       2530 Victoria Court   Milton Mills        ME    3852.0   
499  494 Clarksburg Park Road         Sedona        AZ   86341.0   
500         649 Nutter Street  Overland Park        MO   64110.0   
501  3652 Boone Crockett Lane        Seattle        WA   98109.0   
502         2778 North Avenue           Burr  Nebraska   68324.0   

           country                                          contact  \
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3

In [26]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [28]:
adverse_reactions['adverse_reaction'].value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
cough                         2
throat irritation             2
nausea                        2
Name: adverse_reaction, dtype: int64

In [33]:
# patients from new york
patients[patients['city'] == 'New York']

patient_id assigned_sex    given_name     surname               address  \
9            10       female        Sophie     Cabrera     3303 Anmoore Road   
35           36       female        Kamila    Pecinová  3558 Longview Avenue   
84           85       female         Nương          Vũ   465 Southern Street   
129         130       female       Rebecca    Jephcott      989 Wayback Lane   
142         143         male        Finley    Chandler  2754 Westwood Avenue   
152         153         male   Christopher    Woodward  3450 Southern Street   
188         189         male         Søren    Sørensen      2397 Bell Street   
213         214       female    Onyemaechi   Onwughara     685 Duncan Avenue   
215         216         male          John         Doe       123 Main Street   
229         230         male          John         Doe       123 Main Street   
237         238         male          John         Doe       123 Main Street   
244         245         male          John         Doe       123 Main Street   
247         248         male        Tuukka  Leppäluoto    1886 Bicetown Road   
251         252         male          John         Doe       123 Main Street   
263         264       female         Julia    Carvalho     3662 Shinn Street   
277         278         male          John         Doe       123 Main Street   
301         302       female  Onyekachukwu      Obinna    2970 Forest Avenue   
461         462         male        Cannan     Cabrera   2102 Geraldine Lane   

         city     state  zip_code        country  \
9    New York  New York   10011.0  United States   
35   New York  New York   10004.0  United States   
84   New York        NY   10001.0  United States   
129  New York        NY   10004.0  United States   
142  New York  New York   10001.0  United States   
152  New York        NY   10004.0  United States   
188  New York        NY   10011.0  United States   
213  New York        NY   10013.0  United States   
215  New York        NY   12345.0  United States   
229  New York        NY   12345.0  United States   
237  New York        NY   12345.0  United States   
244  New York        NY   12345.0  United States   
247  New York        NY   10011.0  United States   
251  New York        NY   12345.0  United States   
263  New York        NY   10036.0  United States   
277  New York        NY   12345.0  United States   
301  New York        NY   10004.0  United States   
461  New York        NY   10014.0  United States   

                                               contact   birthdate  weight  \
9        SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7   
35               718-501-0503KamilaPecinova@dayrep.com  12/23/1985   198.9   
84                  VuCamNuong@fleckens.hu516-720-5094    2/1/1981   138.2   
129            631-370-7406RebeccaJephcott@armyspy.com    8/1/1966   203.3   
142              516-740-5280FinleyChandler@dayrep.com  10/25/1936   150.9   
152  ChristopherWoodward@jourrapide.com+1 (516) 630...    9/4/1984   212.2   
188            SrenSrensen@superrito.com1 212 201 3108  12/31/1942   157.1   
213         917-622-9142OnyemaechiOnwughara@einrot.com    3/8/1989   131.1   
215                        johndoe@email.com1234567890    1/1/1975   180.0   
229                        johndoe@email.com1234567890    1/1/1975   180.0   
237                        johndoe@email.com1234567890    1/1/1975   180.0   
244                        johndoe@email.com1234567890    1/1/1975   180.0   
247           917-408-8855TuukkaLeppaluoto@teleworm.us    3/7/1978   211.0   
251                        johndoe@email.com1234567890    1/1/1975   180.0   
263  JuliaAzevedoCarvalho@superrito.com+1 (212) 782...   4/11/1931   171.8   
277                        johndoe@email.com1234567890    1/1/1975   180.0   
301         OnyekachukwuObinna@teleworm.us646-982-6609   1/24/1997   154.7   
461     646-289-4177CannanCabreraOrdonez@superrito.com  10/12/1980   209.7   

     height 

In [36]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [37]:
# 16 Quiz


In [38]:
patients.surname.value_counts()

Doe            6
Jakobsen       3
Taylor         3
Ogochukwu      2
Tucker         2
              ..
Casárez        1
Mata           1
Pospíšil       1
Rukavina       1
Onyekaozulu    1
Name: surname, Length: 466, dtype: int64

In [47]:
patients[patients['surname'] == 'Knudsen']

patient_id assigned_sex given_name  surname address city state  zip_code  \
230         231       female  Elisabeth  Knudsen     NaN  NaN   NaN       NaN   

    country contact  birthdate  weight  height   bmi  
230     NaN     NaN  9/23/1976   165.9      63  29.4

In [40]:
patients.address.value_counts()

123 Main Street             6
2778 North Avenue           2
2476 Fulton Street          2
648 Old Dear Lane           2
3094 Oral Lake Road         1
                           ..
1066 Goosetown Drive        1
4291 Patton Lane            1
4643 Reeves Street          1
174 Lost Creek Road         1
3652 Boone Crockett Lane    1
Name: address, Length: 483, dtype: int64

In [41]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [43]:
patients['weight'].sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [44]:
sum(treatments.auralin.isnull())

0

In [45]:
sum(treatments.novodra.isnull())

0

In [50]:
treatments.sample(10)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
107   flaviano    endrizzi          -  46u - 43u         7.93       7.56   
39   krisztina      magyar          -  32u - 27u         7.52       7.01   
117     javier      moquin          -  30u - 32u         8.00       7.59   
3         skye  gormanston  33u - 36u          -         7.97       7.62   
118    burrell     mathieu  46u - 58u          -         8.71       8.39   
8        saber      ménard          -  54u - 54u         8.08       7.70   
36         chỉ         lâm  45u - 48u          -         7.68       7.24   
189        mee       chung  42u - 56u          -         7.56       7.20   
185       leah        wong          -  28u - 31u         7.79       7.42   
80      hideki   haraguchi          -  37u - 35u         7.59       7.05   

     hba1c_change  
107          0.37  
39           0.51  
117          0.91  
3            0.35  
118          0.32  
8             NaN  
36            NaN  
189          0.36  
185           NaN  
80           0.54

### Tidiness

- Contact colum in patients tables with phone number and email
- 3 variables in 2 columsn in treatment table(tratment, start_dose, end_dose)

In [70]:
patients.columns

Index(['patient_id', 'assigned_sex', 'given_name', 'surname', 'address',
       'city', 'state', 'zip_code', 'country', 'contact', 'birthdate',
       'weight', 'height', 'bmi'],
      dtype='object')

## Cleaning
